pkg/portfwd/client.go

```diff
@@ -52,74 +47,43 @@ func HandleUDPConnection(ctx context.Context, client *guestagentclient.GuestAgen
 		return
 	}
 
-	g, _ := errgroup.WithContext(ctx)
-
-	g.Go(func() error {
-		buf := make([]byte, 65507)
-		for {
-			n, addr, err := conn.ReadFrom(buf)
-			if err != nil {
-				// https://pkg.go.dev/net#PacketConn does not mention io.EOF semantics.
-				if errors.Is(err, io.EOF) {
-					return nil
-				}
-				return err
-			}
-			msg := &api.TunnelMessage{
-				Id:            id + "-" + addr.String(),
-				Protocol:      "udp",
-				GuestAddr:     guestAddr,
-				Data:          buf[:n],
-				UdpTargetAddr: addr.String(),
-			}
-			if err := stream.Send(msg); err != nil {
-				return err
-			}
-		}
-	})
+	//Handshake message to start tunnel
+	err = stream.Send(&api.TunnelMessage{Id: id, Protocol: "udp", GuestAddr: guestAddr})
+	if err != nil {
+		logrus.Errorf("could not start udp tunnel for id: %s error:%v", id, err)
+		return
+	}
 
-	g.Go(func() error {
-		for {
-			in, err := stream.Recv()
-			if err != nil {
-				if errors.Is(err, io.EOF) {
-					return nil
-				}
-				return err
-			}
-			addr, err := net.ResolveUDPAddr("udp", in.UdpTargetAddr)
-			if err != nil {
-				return err
-			}
-			_, err = conn.WriteTo(in.Data, addr)
-			if err != nil {
-				return err
-			}
-		}
+	proxy, err := forwarder.NewUDPProxy(conn, func() (net.Conn, error) {
+		rw := &GrpcClientRW{stream: stream, id: id, addr: guestAddr, protocol: "udp"}
+		return rw, nil
 	})
 
-	if err := g.Wait(); err != nil {
-		logrus.Debugf("error in udp tunnel for id: %s error:%v", id, err)
+	if err != nil {
+		logrus.Errorf("error in udp tunnel proxy for id: %s error:%v", id, err)
+		return
 	}
+
+	defer proxy.Close()
+	proxy.Run()
```

Comment:
We don't get any error, or we ignore it? Are error logged in the proxy? Does ensure termination of both copies on errors or normal completion?

Reply:
`defer proxy.Close()` - Takes care of proper completion. Also we have debug and error logs in UDPProxy

Reply:
Shouldn't you log the error from `Close`?

```
defer func() {
  if err := proxy.Close(); err != nil {
    ...
  }
}()
```

Reply:
If you check the code, proxy.close always returns nil. So handling it i don't think useful

Reply:
Why should we panic ?? 

UDP connect is either way closed. There is no point in sending panic for that case.

Reply:
If you assert that the error *is never nil* then you should write a panic to assert your assumption.

Reply:
Am not interested in doing this change. Since the error during close does not have much value for us i don't see a need for this assert / log.

Additionally from udpProxy code its clear its always nil. So what ever we write its more of a dead code only.

Reply:
Dead code serves as proof evident that the error is guaranteed to be nil.

Reply:
Lets say we get error during close what do you think we need to do ??

Even if we have log how will it help us ???

In tunnels, we open and close lot of connections having these errors in logs adds more of a noise rather than being of any use.

Lets hear what others think. Since we have contradicting opinions. 

@AkihiroSuda what do you think on this ??

Reply:
That's why my position is that you should panic rather than log.

Reply:
To address your concern of this getting missed, Done error log to capture this if in case it happens.

Adding panic in application i don't think its a good idea. Marking this as resolved as err is logged now